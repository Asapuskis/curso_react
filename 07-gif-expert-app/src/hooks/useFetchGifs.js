// Hook custom para hacer la peticiÃ³n de gifs y devolverlos.

import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
            getGifs(category)
                .then( imgs => {
                    // Por defecto, esperaremos 500 milis para poder simular la carga de archivos.
                    setTimeout( () => {
                        setState({
                            data: imgs,
                            loading: false
                        });
                    }, 500);
                });
        },
        [category]
    );

    return state;
}
