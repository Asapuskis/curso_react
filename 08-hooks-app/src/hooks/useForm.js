import { useState } from "react"

// Hook para manejar formularios.
export const useForm = ( initialState = {} ) => {

    const [formValues, setFormValues] = useState(initialState);

    const handleInputChange = ({target}) => {
        setFormValues({
            ...formValues,
            [ target.name ]: target.value
        });
    };

    return {
        formValues,
        handleInputChange
    };

}
