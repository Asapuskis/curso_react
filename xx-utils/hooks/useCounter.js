import { useState } from 'react'

// Este custom hook sirve para manejar el estado de un contador.
// Recuerda que por defecto, un hook empieza por use
export const useCounter = ( initialState = 10 ) => {

    const [state, setState] = useState(initialState)

    const increment = ( factor = 1 ) => {
        setState( state + factor );
    };

    const decrement = ( factor = 1 ) => {
        setState( state - factor);
    }

    const reset = () => {
        setState( initialState );
    }

    return {
        state,
        increment,
        decrement,
        reset
    };
};
