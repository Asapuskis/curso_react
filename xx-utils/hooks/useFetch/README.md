# useFetch Hook

Def:

Hook para hacer llamadas http.

Params:

- url: Direcci贸n para hacer la petici贸n http

Return:

- state: Un objeto con lo siguiente:
  - data: Lo que devuelva la petici贸n http.
  - loading: Si ya ha terminado de cargar la data.
  - error: El mensaje de error en caso de que no se haya podido hacer la petici贸n.

Ejemplo de uso:

```javascript
    const { data, loading, error } = useFetch( url );
```
