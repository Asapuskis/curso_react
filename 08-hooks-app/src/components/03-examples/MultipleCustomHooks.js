import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import './customHooks.css';

export const MultipleCustomHooks = () => {

    const { state:counter, increment } = useCounter(1)
    console.log(counter)

    const { data, loading, error } = useFetch( `https://breakingbadapi.com/api/quotes/${counter}` );

    console.log(data);

    // !! doble negaci√≥n, o sea, si no es null
    // !null == true
    // !!null == false
    // Por lo tanto, si no es null coge la pos 0 de data.
    const { author, quote } = !!data && data[0];

    return (
        <div>
            <h1>Breaking bad quotes!!</h1>
            <hr />
            { loading ?
                <div className='alert alert-info text-center'>
                    Loading...
                </div>
                :
                <blockquote className='blockquote text-right'>
                    <p className='mb-0'>{quote}</p>
                    <footer className='blockquote-footer'>{author}</footer>
                </blockquote>
            }
            <button className='btn btn-primary' onClick={ () => increment(1) }>
                Siguiente quote!
            </button>
        </div>
    )
};
