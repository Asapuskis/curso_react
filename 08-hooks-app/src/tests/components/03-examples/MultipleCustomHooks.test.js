import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');


describe('Pruebas de <MultipleCustomHooks />', () => {

    test('Debe mostrarse un snapshot correctamente', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );
        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe mostrar la info', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'Paco',
                quote: 'Hola'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert').exists()).toBe(false);
        expect( wrapper.find('.mb-0').text().trim()).toBe('Hola');

    });

})