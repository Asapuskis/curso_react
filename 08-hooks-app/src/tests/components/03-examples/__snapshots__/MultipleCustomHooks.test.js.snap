// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <MultipleCustomHooks /> Debe mostrar la info 1`] = `
<div>
  <h1>
    Breaking bad quotes!!
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
      Hola
    </p>
    <footer
      className="blockquote-footer"
    >
      Paco
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote!
  </button>
</div>
`;

exports[`Pruebas de <MultipleCustomHooks /> Debe mostrarse un snapshot correctamente 1`] = `
<div>
  <h1>
    Breaking bad quotes!!
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote!
  </button>
</div>
`;
